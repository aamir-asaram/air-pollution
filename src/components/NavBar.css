nav {
  display: flex;
  justify-content: space-between;
  background-color: rgba(0, 0, 0, 0.1);
  backdrop-filter: blur(10px);
  color: #fff;
  height: max-content;
  position: fixed;
  top: 0;
  width: 100%;
  padding-bottom: 10px;
}

nav > * {
  margin: 1.5rem;
}

nav h2 {
  font-size: 1.5rem;
  font-weight: 300;
  margin-top: 18px;
}

nav button {
  background-color: transparent;
  margin: 0;
  padding: 1rem 0.5rem 0.5rem 0.5rem;
  border: none;
  cursor: pointer;
  font-size: 1.5rem;
  font-weight: 300;
  color: inherit;
}

.color-icon {
  margin-right: 0.5rem;
}
