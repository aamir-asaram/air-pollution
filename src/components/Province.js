import React from 'react';
import PropTypes from 'prop-types';

const categories = [
  'Buffalo City', 'City Of Cape Town', 'City Power', 'Eastern Cape',
  'Eskom Direct', 'Free State', 'Gauteng', 'Kwazulu Natal', 'Limpopo',
  'Mpumalanga', 'North West', 'Northern Cape', 'Western Cape',
];

const baseURL = 'https://eskom-calendar-api.shuttleapp.rs';

const Province = ({ province }) => {
  const fetchDetails = async () => {
    const response = await fetch(`${baseURL}/outages/${province}`);
    const data = await response.json();
    return data;
  };

  let str = province.split('-').join(' ');
  let category;
  categories.forEach((cat) => {
    if (str.includes(cat.toLowerCase())) {
      category = cat;
    }
    str = str.replace(cat.toLowerCase(), '');
  });

  const handleClick = async () => {
    const details = await fetchDetails();
    console.log(details);
  };

  return (
    <li>
      <button type="button" onClick={handleClick}>
        <span className="category">{category}</span>
        <span className="province">{str}</span>
      </button>
    </li>
  );
};

Province.propTypes = {
  province: PropTypes.string.isRequired,
};

export default Province;
