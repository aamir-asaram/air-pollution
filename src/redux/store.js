import { configureStore } from '@reduxjs/toolkit';
import provincesReducer from './provinces/provincesSlice';

const store = configureStore({
  reducer: {
    provinces: provincesReducer,
  },
});

export default store;
