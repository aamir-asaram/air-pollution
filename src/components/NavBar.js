import React from 'react';
import { FaSearch, FaChevronLeft } from 'react-icons/fa';
import './NavBar.css';
import { useNavigate } from 'react-router-dom';

const NavBar = () => {
  const navigate = useNavigate();
  return (
    <nav className="navbar">
      <button
        type="button"
        className="back-btn"
        onClick={() => {
          navigate('/');
        }}
      >
        <FaChevronLeft />
      </button>
      <h2>All Zones</h2>
      <FaSearch className="search-icon" />
    </nav>
  );
};

export default NavBar;
