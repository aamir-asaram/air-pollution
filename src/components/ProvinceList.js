import React, { useEffect, useState } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { fetchProvinces } from '../redux/provinces/provincesSlice';
import Province, { categories } from './Province';
import Current, { fetchDetails } from './Current';
import './provinceList.css';

const ProvinceList = () => {
  const dispatch = useDispatch();
  const provinces = useSelector((state) => state.provinces.provinces);
  const loading = useSelector((state) => state.provinces.loading);
  const error = useSelector((state) => state.provinces.error);
  const [splitProvinces, setSplitProvinces] = useState([]);
  const [filteredProvinces, setFilteredProvinces] = useState([]);

  useEffect(() => {
    if (provinces.length === 0) {
      dispatch(fetchProvinces());
      setFilteredProvinces(provinces);
    } else {
      const updatedSplitProvinces = provinces.map((province) => province.split('-').join(' '));
      setSplitProvinces(updatedSplitProvinces);
      setFilteredProvinces(updatedSplitProvinces);
    }
  }, [dispatch, provinces]);

  if (loading) {
    return <p>Loading...</p>;
  }

  if (error) {
    return <p>Something went wrong...</p>;
  }

  const handleChange = (e) => {
    const filter = e.target.value.toLowerCase();

    if (filter === 'all') {
      setFilteredProvinces(splitProvinces);
    } else {
      const filteredProvinces = splitProvinces.filter(
        (province) => province.includes(filter),
      );
      setFilteredProvinces(filteredProvinces);
    }
  };

  return (
    <div>
      <Current />
      <div id="filter">
        <select onChange={handleChange}>
          <option value="all">All</option>
          {categories.map((category) => (
            <option key={category} value={category}>{category}</option>
          ))}
        </select>
      </div>
      <ul>
        {filteredProvinces.map((province) => (
          <Province key={province} province={province} fetchDetails={fetchDetails} />
        ))}
      </ul>
    </div>
  );
};

export default ProvinceList;
