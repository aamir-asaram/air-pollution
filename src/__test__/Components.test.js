import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import { Provider } from 'react-redux';
import NavBar from '../components/NavBar';
import ProvinceList from '../components/ProvinceList';
import Province from '../components/Province';
import store from '../redux/store';

describe('Test all component rendering', () => {
  it('NavBar renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <NavBar />
          </MemoryRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Province list renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <ProvinceList />
          </MemoryRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Province renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter>
            <Province province="city-of-cape-town" />
          </MemoryRouter>
        </Provider>,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
