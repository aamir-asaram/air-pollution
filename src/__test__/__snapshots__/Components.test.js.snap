// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test all component rendering NavBar renders correctly 1`] = `
<nav
  className="navbar"
>
  <button
    className="back-btn"
    onClick={[Function]}
    type="button"
  >
    <svg
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 320 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M34.52 239.03L228.87 44.69c9.37-9.37 24.57-9.37 33.94 0l22.67 22.67c9.36 9.36 9.37 24.52.04 33.9L131.49 256l154.02 154.75c9.34 9.38 9.32 24.54-.04 33.9l-22.67 22.67c-9.37 9.37-24.57 9.37-33.94 0L34.52 272.97c-9.37-9.37-9.37-24.57 0-33.94z"
      />
    </svg>
  </button>
  <button
    className="color-toggle"
    onClick={[Function]}
    type="button"
  >
    <svg
      className="color-icon"
      fill="currentColor"
      height="1em"
      stroke="currentColor"
      strokeWidth="0"
      style={
        Object {
          "color": undefined,
        }
      }
      viewBox="0 0 512 512"
      width="1em"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M256 160c-52.9 0-96 43.1-96 96s43.1 96 96 96 96-43.1 96-96-43.1-96-96-96zm246.4 80.5l-94.7-47.3 33.5-100.4c4.5-13.6-8.4-26.5-21.9-21.9l-100.4 33.5-47.4-94.8c-6.4-12.8-24.6-12.8-31 0l-47.3 94.7L92.7 70.8c-13.6-4.5-26.5 8.4-21.9 21.9l33.5 100.4-94.7 47.4c-12.8 6.4-12.8 24.6 0 31l94.7 47.3-33.5 100.5c-4.5 13.6 8.4 26.5 21.9 21.9l100.4-33.5 47.3 94.7c6.4 12.8 24.6 12.8 31 0l47.3-94.7 100.4 33.5c13.6 4.5 26.5-8.4 21.9-21.9l-33.5-100.4 94.7-47.3c13-6.5 13-24.7.2-31.1zm-155.9 106c-49.9 49.9-131.1 49.9-181 0-49.9-49.9-49.9-131.1 0-181 49.9-49.9 131.1-49.9 181 0 49.9 49.9 49.9 131.1 0 181z"
      />
    </svg>
  </button>
</nav>
`;

exports[`Test all component rendering Province list renders correctly 1`] = `
<div>
  <div
    id="current"
  >
    <span>
      Current Stage: 
    </span>
    <span>
      0
    </span>
  </div>
  <div
    id="filter"
  >
    <select
      onChange={[Function]}
    >
      <option
        value="all"
      >
        All
      </option>
      <option
        value="Buffalo City"
      >
        Buffalo City
      </option>
      <option
        value="City Of Cape Town"
      >
        City Of Cape Town
      </option>
      <option
        value="City Power"
      >
        City Power
      </option>
      <option
        value="Eastern Cape"
      >
        Eastern Cape
      </option>
      <option
        value="Eskom Direct"
      >
        Eskom Direct
      </option>
      <option
        value="Free State"
      >
        Free State
      </option>
      <option
        value="Gauteng"
      >
        Gauteng
      </option>
      <option
        value="Kwazulu Natal"
      >
        Kwazulu Natal
      </option>
      <option
        value="Limpopo"
      >
        Limpopo
      </option>
      <option
        value="Mpumalanga"
      >
        Mpumalanga
      </option>
      <option
        value="North West"
      >
        North West
      </option>
      <option
        value="Northern Cape"
      >
        Northern Cape
      </option>
      <option
        value="Western Cape"
      >
        Western Cape
      </option>
    </select>
  </div>
  <ul />
</div>
`;

exports[`Test all component rendering Province renders correctly 1`] = `
<li>
  <button
    onClick={[Function]}
    type="button"
  >
    <span
      className="category"
    >
      City Of Cape Town
    </span>
    <span
      className="province"
    />
  </button>
</li>
`;
